src/gateway/environments/jsdom-sixteen/jsdom-sixteen-resource-loader.js

```diff
@@ -0,0 +1,144 @@
+// @flow
+import {URL} from "url";
+import {ResourceLoader} from "jsdom";
+import type {FetchOptions} from "jsdom";
+import type {RequestOptions, RenderAPI} from "../../types.js";
+import {getAgentForURL} from "../../../shared/index.js";
+import {
+    DefaultRequestOptions,
+    request,
+    abortInFlightRequests,
+} from "../../request.js";
+import {applyAbortablePromisesPatch} from "./apply-abortable-promises-patch.js";
+
+/**
+ * A ResourceLoader implementation for JSDOM sixteen-compatible versions of
+ * JSDOM that only allows for fetching JS files.
+ *
+ * A JS file request is identified by the regular expression:
+ *   /^.*\.js(?:\?.*)?/g
+ */
+export class JSDOMSixteenResourceLoader extends ResourceLoader {
+    /**
+     * Used to indicate if any pending requests are still needed so that we
+     * can report when an unused request is fulfilled.
+     */
+    _active: boolean;
+    _renderAPI: RenderAPI;
+    _requestOptions: RequestOptions;
+
+    static get EMPTY_RESPONSE(): Promise<Buffer> {
+        return Promise.resolve(Buffer.from(""));
+    }
+
+    /**
+     * Create instance of the resource loader.
+     *
+     * @param {RequestFn} requestFn
+     * The function responsibly for fulfilling GET requests for URLs.
+     */
+    constructor(
+        renderAPI: RenderAPI,
+        requestOptions?: RequestOptions = DefaultRequestOptions,
+    ) {
+        // Patch before super to make sure promises get an abort.
+        applyAbortablePromisesPatch();
```

Comment:
We make sure that the patch is applied here.

Reply:
Do we think that this is something we'll need to do for every environment we have? Would it be worth abstracting this out?

Reply:
I don't think we will need to, no. This is specifically a hack for JSDOM and one that we soon won't need as they just fixed the underlying issue about 2 days ago.